master/buildbot/reporters/notifier.py

```diff
@@ -55,85 +48,104 @@ def checkConfig(self, mode=("failing", "passing", "warnings"),
                         subject="Buildbot %(result)s in %(title)s on %(builder)s",
                         addLogs=False, addPatch=False,
                         schedulers=None, branches=None,
-                        watchedWorkers=None, messageFormatterMissingWorker=None):
-
-        self.mode = self.computeShortcutModes(mode)
-        self.tags = tags
-        self.builders = builders
-        self.schedulers = schedulers
-        self.branches = branches
-        self.subject = subject
-        self.addLogs = addLogs
-        self.addPatch = addPatch
-        if messageFormatter is None:
-            messageFormatter = DefaultMessageFormatter()
-        self.messageFormatter = messageFormatter
-        if messageFormatterMissingWorker is None:
-            messageFormatterMissingWorker = MessageFormatterMissingWorker()
-        self.messageFormatterMissingWorker = messageFormatterMissingWorker
-        self.buildSetSummary = buildSetSummary
-        self._buildset_complete_consumer = None
-        if watchedWorkers is None:
-            self.watchedWorkers = ()
-        else:
-            self.watchedWorkers = watchedWorkers
+                        watchedWorkers=None, messageFormatterMissingWorker=None,
+                        generators=None):
+
+        if generators is None:
+            generators = self.create_generators_from_old_args(mode, tags, builders, buildSetSummary,
+                                                              messageFormatter, subject, addLogs,
+                                                              addPatch, schedulers,
+                                                              branches, watchedWorkers,
+                                                              messageFormatterMissingWorker)
+
+        self.generators = generators
+        self._needs_buildsets_complete = \
+            any(isinstance(g, BuildSetStatusGenerator) for g in generators)
+        self._needs_build_finished = \
+            any(isinstance(g, BuildStatusGenerator) for g in generators)
+        self._needs_worker_missing = \
+            any(isinstance(g, WorkerMissingGenerator) for g in generators)
 
     @defer.inlineCallbacks
     def startService(self):
         yield super().startService()
         startConsuming = self.master.mq.startConsuming
-        self._buildsetCompleteConsumer = yield startConsuming(
-            self.buildsetComplete,
-            ('buildsets', None, 'complete'))
-        self._buildCompleteConsumer = yield startConsuming(
-            self.buildComplete,
-            ('builds', None, 'finished'))
-        self._workerMissingConsumer = yield startConsuming(
-            self.workerMissing,
-            ('workers', None, 'missing'))
 
-    @defer.inlineCallbacks
+        self._buildsetCompleteConsumer = None
+        self._buildCompleteConsumer = None
+        self._workerMissingConsumer = None
+
+        if self._needs_buildsets_complete:
```

Comment:
This cannot work in the reconfig use case. The flag would be updated but the consumer will not be stopped or created according to config change.

Reply:
Fixed now.